<a href="https://colab.research.google.com/github/tjwei/DataScienceSoftware2026/blob/main/pandas01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd

In [3]:
pd.read_html('https://en.wikipedia.org/wiki/List_of_countries_and_dependencies_by_population')

HTTPError: HTTP Error 403: Forbidden

In [4]:
!curl -JLO https://en.wikipedia.org/wiki/List_of_countries_and_dependencies_by_population

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1351k  100 1351k    0     0  6650k      0 --:--:-- --:--:-- --:--:-- 6656k


In [6]:
tables = pd.read_html('List_of_countries_and_dependencies_by_population')
len(tables)

3

### 解決 403 錯誤：加入 User-Agent
透過 `storage_options` 參數，我們可以自訂 HTTP 標頭（Headers），讓網站認為這是一般瀏覽器發出的請求。

In [27]:
import pandas as pd

url = 'https://en.wikipedia.org/wiki/List_of_countries_and_dependencies_by_population'

headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'}
tables = pd.read_html(url, storage_options=headers)

print(len(tables))

df = tables[0]
display(df.head())

3


,Location,Population,% of world,Date,Source (official or from the United Nations),Notes
0,World,8.232000e+09,100%,13 Jun 2025,UN projection[1][3],NaN
1,India,1.429404e+09,17.3%,1 Jul 2026,Official projection[4],[b]
2,China,1.404890e+09,17.0%,31 Dec 2025,Official estimate[5],[c]
3,United States,3.417849e+08,4.1%,1 Jul 2025,Official estimate[6],[d]
4,Indonesia,2.883151e+08,3.5%,31 Dec 2025,National annual projection[7],NaN


In [28]:
df.shape

(241, 6)

In [29]:
df.dtypes

,0
Location,object
Population,float64
% of world,object
Date,object
Source (official or from the United Nations),object
Notes,object


In [30]:
df

,Location,Population,% of world,Date,Source (official or from the United Nations),Notes
0,World,8.232000e+09,100%,13 Jun 2025,UN projection[1][3],NaN
1,India,1.429404e+09,17.3%,1 Jul 2026,Official projection[4],[b]
2,China,1.404890e+09,17.0%,31 Dec 2025,Official estimate[5],[c]
3,United States,3.417849e+08,4.1%,1 Jul 2025,Official estimate[6],[d]
4,Indonesia,2.883151e+08,3.5%,31 Dec 2025,National annual projection[7],NaN
...,...,...,...,...,...,...
236,Christmas Island (Australia),1.692000e+03,0%,1 Jan 2021,2021 Census[247],NaN
237,Niue (New Zealand),1.681000e+03,0%,11 Nov 2022,2022 Census[248],NaN
238,Vatican City,8.820000e+02,0%,31 Dec 2024,Official figure[249],[ah]
239,Cocos (Keeling) Islands (Australia),5.930000e+02,0%,30 Jun 2020,2021 Census[250],NaN


In [32]:
df = df.dropna(how='all')

In [33]:
df.shape

(240, 6)

In [43]:
df[df['Location'].str.startswith('Ta')]

,Location,Population,% of world,Date,Source (official or from the United Nations),Notes
21,Tanzania,70038850.0,0.8%,1 Jul 2026,Official projection[25],[h]
60,Taiwan,23224721.0,0.3%,31 Aug 2026,Official estimate[72],[p]
89,Tajikistan,10721000.0,0.1%,1 Jan 2026,Official estimate[100],NaN


In [53]:
df.loc[1:].Population.mean()

np.float64(33783539.033472806)

In [55]:
df[1:].Population.quantile([0.25,0.5,0.75, 0.8])

,Population
0.25,382138.0
0.50,5557096.0
0.75,23113967.5
0.80,32669187.4


In [62]:
df[df.Population > 100_000_000][1:].Population.sum()/df.loc[0].Population

np.float64(0.639882451287658)

In [65]:
df.loc[0].Population

np.float64(8232000000.0)

In [68]:
df.Population[0]

np.float64(8232000000.0)

In [71]:
df.loc[0]

,0
Location,World
Population,8232000000.0
% of world,100%
Date,13 Jun 2025
Source (official or from the United Nations),UN projection[1][3]
Notes,NaN


In [76]:
df['Population']

,Population
0,8.232000e+09
1,1.429404e+09
2,1.404890e+09
3,3.417849e+08
4,2.883151e+08
...,...
236,1.692000e+03
237,1.681000e+03
238,8.820000e+02
239,5.930000e+02


In [78]:
df.T

,0,1,2,3,4,5,6,7,8,9,...,230,231,232,233,234,236,237,238,239,240
Location,World,India,China,United States,Indonesia,Pakistan,Nigeria,Brazil,Bangladesh,Russia,...,"Saint Helena, Ascension and Tristan da Cunha (UK)",Montserrat (UK),Falkland Islands (UK),Tokelau (New Zealand),Norfolk Island (Australia),Christmas Island (Australia),Niue (New Zealand),Vatican City,Cocos (Keeling) Islands (Australia),Pitcairn Islands (UK)
Population,8232000000.0,1429404000.0,1404890000.0,341784857.0,288315089.0,241499431.0,223800000.0,214211951.0,169828911.0,146028325.0,...,5651.0,4386.0,3662.0,2608.0,2188.0,1692.0,1681.0,882.0,593.0,35.0
% of world,100%,17.3%,17.0%,4.1%,3.5%,2.9%,2.7%,2.6%,2.1%,1.8%,...,0%,0%,0%,0%,0%,0%,0%,0%,0%,0%
Date,13 Jun 2025,1 Jul 2026,31 Dec 2025,1 Jul 2025,31 Dec 2025,1 Mar 2023,1 Jul 2023,1 Jul 2026,14 Jun 2022,1 Jan 2025,...,1 Jul 2021,23 Sep 2023,10 Oct 2021,1 Jan 2025,1 Jan 2021,1 Jan 2021,11 Nov 2022,31 Dec 2024,30 Jun 2020,2023
Source (official or from the United Nations),UN projection[1][3],Official projection[4],Official estimate[5],Official estimate[6],National annual projection[7],2023 census result[8],Official projection[9],Official estimate[10],2022 census result[11],Official estimate[13],...,2021 Census[240][241][242],2023 census result[243],2021 census result[244],2025 Census[245],2021 Census[246],2021 Census[247],2022 Census[248],Official figure[249],2021 Census[250],Official estimate[251]
Notes,NaN,[b],[c],[d],NaN,[e],NaN,NaN,[f],[g],...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,[ah],NaN,NaN


In [85]:
(df.T)[236]

,236
Location,Christmas Island (Australia)
Population,1692.0
% of world,0%
Date,1 Jan 2021
Source (official or from the United Nations),2021 Census[247]
Notes,NaN


In [81]:
df[0]

KeyError: 0